File: Cpp_src/medium/remove_nth_node_from_end_of_list_19.cc
#include <iostream>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// Use Two pass algorithm
// Time Complexity: O(L)
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 86.13%
// Memory Usage: 8.5 MB, less than 76.32%
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        if (head->next == NULL)
        {
            return NULL;
        }
        ListNode *cur = head;
        int length = 0;

        while (cur)
        {
            length++;
            cur = cur->next;
        }
        if (n == length)
        {
            return head->next;
        }

        cur = head;
        for (int i = 0; i < (length - n - 1); i++)
        {
            cur = cur->next;
        }
        ListNode *tmp = cur->next;
        cur->next = cur->next->next;
        delete tmp;

        return head;
    }
};

// Use One pass algorithm
// Time Complexity: O(L)
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 86.13%
// Memory Usage: 8.5 MB, less than 80.26%
class Solution2
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        if (head->next == NULL)
        {
            return NULL;
        }
        ListNode *fast = head;
        ListNode *slow = head;

        while (n--)
        {
            fast = fast->next;
        }
        if (fast == NULL)
        {
            return head->next;
        }
        while (fast->next)
        {
            slow = slow->next;
            fast = fast->next;
        }
        ListNode *tmp = slow->next;
        slow->next = slow->next->next;
        delete tmp;

        return head;
    }
};

File: Golang_src/medium/removeNthNodeFromEndOfList_19.go
package medium

// Use two pointers:
// Time Complexity: O(L)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.2 MB, less than 21.37%
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head.Next == nil {
		return nil
	}

	slow, fast := head, head
	for n != 0 {
		fast = fast.Next
		n--
	}
	if fast == nil {
		return head.Next
	}

	for fast.Next != nil {
		fast = fast.Next
		slow = slow.Next
	}

	slow.Next = slow.Next.Next

	return head
}

// Use two pointers enhance version:
// Time Complexity: O(L)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.2 MB, less than 21.37%
func removeNthFromEnd2(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	slow, fast := dummy, head
	for i := 0; i < n; i++ {
		fast = fast.Next
	}
	for ; fast != nil; fast = fast.Next {
		slow = slow.Next
	}
	slow.Next = slow.Next.Next

	return dummy.Next
}
